### PR TITLE
Reduce duplications and complexity in v3.0 JSON Schema

### PR DESCRIPTION
**Pull request by vearutop**
Current [WIP Schema](https://github.com/OAI/OpenAPI-Specification/pull/1270) suffers from copy-paste, with these changes it can be improved.

## Combine HTTP Methods in `patternProperties`

```yaml
patternProperties:
  '^(get|put|post|delete|options|head|patch|trace)$': $ref: '#/definitions/Operation'
```

## Combine some definitions as filtered supersets

In order to avoid massive duplication all possible properties can be defined in a superset.
Custom rules of exclusiveness can be further defined as a list of traits in `allOf`.

For example such structure does not allow having `example` and `examples` in same object:
```yaml
not:
  required:
    - example
    - examples
```

## Embed references

In order to simplify `oneOf` logic, many references can be embedded in respective definitions 

All references to `#/definitions/Schema` are happening in such form:

```yaml
oneOf:
  - $ref: '#/definitions/Reference'
  - $ref: '#/definitions/Schema'
```

Therefore `#/definitions/Reference` can be merged into `#/definitions/Schema` to simplify usages:

```yaml
$ref: '#/definitions/Schema'
```

Validation difference after this change is that additional unknown properties are no longer accepted for References.
Such data would become invalid:

```json
{"$ref": "#", "unknown":  1}
```

## Validate example schemas

Modified schema fails validation against example schemas:

* `petstore-expanded.yaml`: `GET /pets` has `query` parameter `tags` with invalid style `simple`,
* `api-with-examples.yaml`: Object expected for `examples`, array received.

### CODE REVIEW

**Comment by handrews** [comment]
@vearutop I wish you'd commented on the issue before doing this.  I'm thrilled to have someone else work on it but this is mostly but not quite a duplicate of what I was about to post later this week- it only didn't get posted during December because I got really sick for several weeks.  So now I guess we try to put PRs next to each other and figure out the overlap and what needs to be kept from each :-/

**Comment by handrews** [comment]
@vearutop if you have any more of these in the works, please let me know- as I noted, it's great that you're working on this, and you may get things done faster than me.  I'd just like to avoid duplication.

**Comment by vearutop** [comment]
I did this refactoring in just couple of hours, more like a demo of alternative approach. I don't mind to withdraw my PR or rebase it on anything else.

**Comment by handrews** [comment]
@vearutop oh, cool, glad this was a quick thing for you- I'll take a look in more detail.  I'm more than happy to recommend merging yours if it works.  I am not possessive here :-)

**Comment by handrews** [comment]
@vearutop @webron embarrassingly since commenting above I've been plagued by a series of minor illnesses and never quite getting caught up enough to attend to the OAS schema.  :-(  I haven't given up, but I'm currently coming down with my 4th cold in 2.5 months and I'm not sure when I'll be back to full strength and caught up with all of my projects.

File: schemas/v3.0/schema.yaml [hunk 1]
```diff
@@ -756,540 +617,164 @@ definitions:
...
-      - in
-      - schema
+      - scheme
+      - type
     properties:
-      name:
+      scheme:
         type: string
-      in:
+        not:
```
Comment (handrews) [inline schemas/v3.0/schema.yaml]: I'm not sure that your combinations of `not` and `enum` all work out.  

Much farther down, you `oneOf` both bearer and not bearer, but you have not bearer here.  I think you just want `type: string` here without any enums, and only do the enum in the `oneOf`

There are some other `enum`/`oneOf` combinations that look similar that may be worth a closer look.  I have not gone through all in detail.

Comment (vearutop) [reply schemas/v3.0/schema.yaml]: Indeed, this `not enum` should not be there, thank you! I checked other places with `not` or `enum` and they seem to be ok. 

But the best would be to run validation against some large set of OAI schema to see if there is a regression.

**Comment by MikeRalphson** [comment]
@vearutop could you look at breaking this PR down into smaller commits? The change I've so far spotted problems with is with folding the `$ref` object definition into surrounding objects and removing the `oneOf`s. If you look specifically at the `response` object, you seem to have made `$ref` a pattern property of the main object, but the constraint that `description` is `required` is still present. This leads my test suite to fail.

**Comment by vearutop** [comment]
Sure, I can do that. Could you share few schemas that are failing validation if they are in public domain?

**Comment by MikeRalphson** [comment]
Sure, all of these point into https://github.com/APIs-guru/openapi-directory

```
../openapi-directory/APIs/6-dot-authentiqio.appspot.com/6/swagger.yaml:
../openapi-directory/APIs/adafruit.com/2.0.0/swagger.yaml:
../openapi-directory/APIs/agco-ats.com/v1/swagger.yaml:
../openapi-directory/APIs/aiception.com/1.0.0/swagger.yaml:
../openapi-directory/APIs/amazonaws.com/AWSMigrationHub/2017-05-31/swagger.yaml:
../openapi-directory/APIs/amazonaws.com/acm-pca/2017-08-22/swagger.yaml:
```

**Comment by vearutop** [comment]
@MikeRalphson indeed I underestimated the effect of embedding references. Having `$ref` in `patternProperties` of respective schema affects `required` and `oneOf`/`allOf`. Now I think the benefits of having embedded refs are not enough so I reverted that commit.

Please check again.

**Comment by MikeRalphson** [comment]
No other regressions detected across the APIs-guru collection. Good job!

**Comment by webron** [comment]
@vearutop we're having a meeting dedicated to the JSON Schema tomorrow. Feel free to join us if you want. Details are at https://openapi.groups.io/g/tsc/viewevent?repeatid=3837&eventid=428242&calstart=2019-03-21.

**Comment by darrelmiller** [comment]
Summary of TSC discussion
- We would like to revert the references to schema objects to use the oneOf syntax as referenced above in the "Embed Reference" comment.

**Comment by vearutop** [comment]
@darrelmiller done, please check.

**Comment by webron** [comment]
Thanks for all the work, @vearutop! You've definitely managed to elegantly reduce the size of the JSON Schema without compromising validation. Top-notch work!

@OAI/tsc - if I can get one or two more approvals, I'll merge this and make the final additions to my PR.

**Pull request merged**